crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -98,12 +98,32 @@ fn runtime_required_decorators(
     }
 
     decorator_list.iter().any(|decorator| {
+        let expression = map_callable(&decorator.expression);
         semantic
-            .resolve_qualified_name(map_callable(&decorator.expression))
+            .resolve_qualified_name(expression)
+            // if we can't resolve the name, then try resolving the assignment
+            .or_else(|| {
+                let mut source = expression;
+                let mut tail = vec![];
+                while let Expr::Attribute(ast::ExprAttribute { value, attr, .. }) = source {
```

Comment:
Could we use `semantic.lookup_attribute`?

Reply:
I would've liked to, but it only gives us the fully qualified name, but we also need the name node at the bottom, so we can `resolve_assignment` on it.

I also considered shifting this complexity into `resolve_assignment`, or adding a new function to `analyze::typing`, but that wouldn't really help us reuse `lookup_attribute` either.

Reply:
Yeah, our semantic model API is a bit annoying here. 

So, the "cheap" solution here would be to do the same as `is_logger_candidate` where we match on the `UnqualifiedName::from_expr` directly. 

Or we keep the "clever" solution that tries to follow at least one alias. This makes me wonder if this logic should be moved into `resolve_assignment`. I think it would make the API more intuitive overall because it isn't very intuitive that the first argument must be a name and we have multiple rules now that manually unroll one level of attributes. 

I think it could help performance to use a `SmallVec` with at least a size of 3 to avoid allocating in the common case.

Reply:
We could generate a faux-`QualifiedName` when `resolve_qualified_name` fails from `UnqualifiedName` prefixed with the current module (using the same logic as `resolve_qualified_name` uses for class/function definitions)

That would get rid of the need of having to add two entries to support both the same/foreign module use-case, i.e. in the test you'd only need `module.app.app.get` and not `fastapi.FastAPI.get` as well. The only drawback would be that you'd need an entry for each app instance. But you would need that anyways if you share the instances across module boundaries.

The only part of that, which feels a bit icky is, that red knot will be able to infer the fully qualified name even across module boundaries. So eventually the correct entry will be `fastapi.FastAPI.get` rather than `module.app.app.get`.

Reply:
I could see us returning a `Result` from `resolve_qualified_name` where the `Err` returns an `UnqualifiedName`, but I don't think we should make up qualified names; this seems somewhat error-prone. But I think we should still try to resolve the `UnqualifiedName`, e.g if you have:

```
import fastapi

a = fastapi

a.FastAPI # the full qualified name should resolve to `fastapi.FastAPI`
```

Which is what `resolve_assignment` does today, at least if you call it with `a` (but not if called with `a.FastAPI`)

Reply:
Yes, that's why it feels icky to do that, but it would solve the issue of having redundant entries.

Although your proposed generic solution for an additional setting to map qualified names seems more robust. Especially once we transition to red knot as the backend where we no longer need to do that mapping manually.

The only problem with the generic solution is that we either end up adding the same translation call after all of the `resolve_qualified_name` calls that benefit from this translation or we build it into `resolve_qualified_name` itself and end up with a potentially fairly big overhead if the list of mappings is large.

Reply:
We might be able to reduce the overhead by, e.g., using [`matchit`](https://docs.rs/matchit/latest/matchit/) similar to what we do when resolving configurations, but I haven't fully thought that through. I'd say the good thing is the list will be small for most cases, so it might not be as big of a perf problem.

But I think we can tackle the extra setting separately. What we need in this PR is to resolve to `module.app.App` so that it can later be rewritten to `fastapi.FastAPI`

Reply:
Yup I agree. I will switch to `SmallVec`, move the attribute logic to `resolve_assignment` and clean up any call sites that do the manual 1-wide attribute extension. Unless you want to do that refactor in a separate PR as well.

Reply:
I think doing it in this PR is fine. There are only five call-sites but it's up to you

Reply:
Turns out the other call sites don't really need to change, the only one that might slightly benefit is the airflow rule, although it would make the logic for the replacement more complicated, so I left it as is for now.